File: day_5/puzzle_1/solve.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const benchmark_1 = require("../../benchmark/benchmark");
(0, benchmark_1.benchmark)(() => {
    const filePath = path_1.default.resolve(__dirname, "input.txt");
    const file = fs_1.default.readFileSync(filePath, { encoding: "utf-8" });
    const [stacks, instructions] = file.split("\n\n");
    const stackGrid = stacks
        .split("\n")
        .reverse()
        .reduce((grid, row, i) => {
        if (i === 0) {
            return grid;
        }
        row.split("").forEach((char, j) => {
            if (char === " ") {
                return;
            }
            if ((j - 1) % 4 !== 0) {
                return;
            }
            const position = (j - 1) / 4;
            grid[position].push(char);
        });
        return grid;
    }, [[], [], [], [], [], [], [], [], []]);
    instructions
        .replaceAll(/\b(move |from |to )\b/g, "")
        .split("\n")
        .forEach((instruction) => {
        const [move, from, to] = instruction.split(" ").map(Number);
        for (let i = 0; i < move; i++) {
            const crate = stackGrid[from - 1].pop();
            if (!crate) {
                throw new Error("🤔");
            }
            stackGrid[to - 1].push(crate);
        }
    });
    const topCrates = stackGrid.reduce((string, col) => `${string}${col.at(-1)}`, "");
    return topCrates;
});

File: day_5/puzzle_1/solve.py
from functools import reduce
import os
import re
from typing import List

from benchmark import benchmark


def main():
    directory = os.path.dirname(os.path.abspath(__file__))
    file_path = os.path.join(directory, "input.txt")

    with open(file_path, "r") as file:
        stacks, instructions = file.read().split("\n\n")

        stack_lines = reversed(stacks.splitlines())

        stack_grid = [[] for _ in range(9)]

        for i, row in enumerate(stack_lines):
            if i == 0:
                continue

            for j, char in enumerate(row):
                if char == " ":
                    continue
                if (j - 1) % 4 != 0:
                    continue

                position = (j - 1) // 4

                stack_grid[position].append(char)

        processed_instructions = re.sub(
            r"\b(move |from |to )\b", "", instructions
        ).splitlines()

        for instruction in processed_instructions:
            move, fromm, to = map(int, instruction.split(" "))

            for _ in range(move):
                crate = stack_grid[fromm - 1].pop()
                stack_grid[to - 1].append(crate)

        top_crates = "".join(col[-1] for col in stack_grid)

        return top_crates


if __name__ == "__main__":
    benchmark(main)
